Use `filepath.Clean` in `normaliseWorkdir` for windows

yongtang opened the pull request:
**\- What I did**

As is seem in the comment of `normaliseWorkdir` for windows, the behavior of `normaliseWorkdir` is supposed to be:

```
...
    //  WORKDIR c:\\foo                 --> C:\foo
    //  WORKDIR \\foo                   --> C:\foo
...
```

However, this is not the case in the current implementation because `filepath.FromSlash` is used and `FromSlash` does not replace multiple separator with a single one (`file.Clean` does).
So `normaliseWorkdir` does not truly normalize workdir.

**\- How I did it**

This fix changes the implementation of `normaliseWorkdir` and use `filepath.Clean` instead of `filepath.FromSlash`.

**\- How to verify it**

Additional test cases have been added to the unit test.

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

Signed-off-by: Yong Tang yong.tang.github@outlook.com

cpuguy83 commented:
Not sure if it's expected to be supported, but filepath.Clean turns `c:` into `c:.`.
Would be worth adding a test case at least.

ping @jhowardmsft

jhowardmsft commented:
I think this is OK. I was initially concerned that filepath.Clean didn't `FromSlash` the result, but it appears (at least in 1.7) the implementation does.

yongtang commented:
Thanks @jhowardmsft @cpuguy83 for the review.

I take a look at the case of `C:`. It seems that on Windows, `C:` is considered as the `C:\.` (**Current directory** on `C:`) so it might actually referring to `C:\Users\myuser`
So technically, `C:` should not be considered as an absolute path unless we resolve the `C:` to the actually path e.g., `C:\Users\myuser` etc. 

I am not sure what is the best way to handle cases of `C:` and `C:\.`. Should we resolve the path into the actually path (`C:\Users\myuser`) and pass it to the workdir?

jhowardmsft commented:
Hmmm. Yes, archaic hangover from DOS days. Personally, I'd prefer to treat anything that has to be resolved to `whatever the current directory on drive 'X:' is' as an error case. If there's significant feedback that this is a problem, it could be possibly addressed in the future (if indeed it's possible - it may not be in the case of Hyper-V Containers).

cpuguy83 commented:
For volumes we are treating `c:` the same as `c:\`

thaJeztah commented:
What's the status here? Do we need additional test cases for the things mentioned above? Or is this good to go as-is?

yongtang committed:
Use `filepath.Clean` in `normaliseWorkdir` for windows

As is seem in the comment of `normaliseWorkdir` for windows:
```
...
	//	WORKDIR c:\\foo                 --> C:\foo
	//	WORKDIR \\foo                   --> C:\foo
...
```

However, this is not the case in the current implementation because
`filepath.FromSlash` is used and `FromSlash` does not replace multiple
separator with a single one (`file.Clean` does).
So `normaliseWorkdir` does not truly normalize workdir.

This fix changes the implementation of `normaliseWorkdir` and use
`filepath.Clean` instead of `filepath.FromSlash`.

Additional test cases have been added to the unit test.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

jhowardmsft commented:
In the case of volumes though, `c:` as a destination drive (if that's what you mean) is correct and is effectively `c:\`. Bad example though, `d:` would be better. I'm just not sure that it is here in the case of `WORKDIR`. As I mentioned above, I think the right thing to do is to error here, and have test cases that validate that.

yongtang commented:
Thanks @jhowardmsft @thaJeztah @cpuguy83. The PR has been updated which treat `C:.` as an error and several test cases have been added. Please take a look and let me know if there are any issues.

cpuguy83 commented on builder/dockerfile/dispatchers_windows.go:
```diff
@@ -17,8 +17,27 @@ func normaliseWorkdir(current string, requested string) (string, error) {
+	// Since filepath.Clean() will automatically normalize the above
+	// to `C:.`, we only need to check the last format
+	if len(current) == 3 && strings.HasSuffix(current, ":.") {
+		return "", fmt.Errorf("cannot use indefinitive current directory '%s'", current)
```
`"%s is not a directory. If you are specifying a driver letter, please add a trailing '\'"`

yongtang replied:
Thanks @cpuguy83. The PR has been updated. Please take a look and let me know if there are any issues.

yongtang commented:
The failure in janky is unrelated (see #26588):
```
21:28:50 ----------------------------------------------------------------------
21:28:50 FAIL: docker_cli_swarm_test.go:246: DockerSwarmSuite.TestSwarmContainerAutoStart
21:28:50 
21:28:50 [ddddd93a387a3] waiting for daemon to start
21:28:50 [ddddd93a387a3] daemon started
21:28:50 [ddddd93a387a3] exiting daemon
21:28:50 [ddddd93a387a3] waiting for daemon to start
21:28:50 [ddddd93a387a3] daemon started
21:28:50 docker_cli_swarm_test.go:265:
21:28:50     c.Assert(strings.TrimSpace(out), checker.Not(checker.Equals), "")
21:28:50 ... obtained string = ""
21:28:50 ... expected string = ""
21:28:50 
21:28:50 [ddddd93a387a3] exiting daemon
21:28:52 
21:28:52 
```

vdemeester commented:
ping @cpuguy83 @jhowardmsft

jhowardmsft commented on builder/dockerfile/dispatchers_windows.go:
```diff
@@ -17,8 +17,27 @@ func normaliseWorkdir(current string, requested string) (string, error) {
+	// C:
+	// C:.
+	// then an error will be thrown as the definition for the above
+	// refers to `current directory on driver C:`
```
typo: `drive`

jhowardmsft commented on builder/dockerfile/dispatchers_windows.go:
```diff
@@ -17,8 +17,27 @@ func normaliseWorkdir(current string, requested string) (string, error) {
+	// Since filepath.Clean() will automatically normalize the above
+	// to `C:.`, we only need to check the last format
+	if len(current) == 3 && strings.HasSuffix(current, ":.") {
+		return "", fmt.Errorf("%s is not a directory. If you are specifying a driver letter, please add a trailing '\\'", current)
```
typo `drive`

jhowardmsft commented on builder/dockerfile/dispatchers_windows.go:
```diff
@@ -17,8 +17,27 @@ func normaliseWorkdir(current string, requested string) (string, error) {
+		return "", fmt.Errorf("%s is not a directory. If you are specifying a driver letter, please add a trailing '\\'", current)
+	}
+	if len(requested) == 3 && strings.HasSuffix(requested, ":.") {
+		return "", fmt.Errorf("%s is not a directory. If you are specifying a driver letter, please add a trailing '\\'", requested)
```
typo `drive`

jhowardmsft commented on builder/dockerfile/dispatchers_windows.go:
```diff
@@ -17,8 +17,27 @@ func normaliseWorkdir(current string, requested string) (string, error) {
+	// refers to `current directory on driver C:`
+	// Since filepath.Clean() will automatically normalize the above
+	// to `C:.`, we only need to check the last format
+	if len(current) == 3 && strings.HasSuffix(current, ":.") {
```
There should be a further check that the first character (lower-cased) is `a..z`

jhowardmsft commented on builder/dockerfile/dispatchers_windows.go:
```diff
@@ -17,8 +17,27 @@ func normaliseWorkdir(current string, requested string) (string, error) {
+	if len(current) == 3 && strings.HasSuffix(current, ":.") {
+		return "", fmt.Errorf("%s is not a directory. If you are specifying a driver letter, please add a trailing '\\'", current)
+	}
+	if len(requested) == 3 && strings.HasSuffix(requested, ":.") {
```
Same comment about first character

yongtang commented:
Thanks @jhowardmsft for the review. The PR has been updated. Please take a look and let me know if there are any other issues.

jhowardmsft commented:
Thanks @yongtang. LGTM if passes Jenkins.

tiborvass commented on builder/dockerfile/dispatchers_windows.go:
```diff
@@ -17,8 +17,28 @@ func normaliseWorkdir(current string, requested string) (string, error) {
+	// refers to `current directory on drive C:`
+	// Since filepath.Clean() will automatically normalize the above
+	// to `C:.`, we only need to check the last format
+	pattern := regexp.MustCompile(`^[a-zA-Z]:\.$`)
```
this variable can be in a top-level `var` declaration.

yongtang replied:
Thanks @tiborvass. The PR has been updated. Please let me know if there are any issues.

tiborvass commented:
LGTM
